// gatsby-cli/lib/reporter/reporter-progress.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.createProgressReporter = void 0;

var _actions = _interopRequireDefault(require("./redux/actions"));

var _constants = require("./constants");

const createProgressReporter = ({
  id,
  text,
  start,
  total,
  span,
  reporter
}) => {
  let lastUpdateTime = 0;
  let unflushedProgress = 0;
  let unflushedTotal = 0;
  const progressUpdateDelay = Math.round(1000 / 10); // 10 fps *shrug*

  const updateProgress = (forced = false) => {
    const t = Date.now();
    if (!forced && t - lastUpdateTime <= progressUpdateDelay) return;

    if (unflushedTotal > 0) {
      _actions.default.setActivityTotal({
        id,
        total: unflushedTotal
      });

      unflushedTotal = 0;
    }

    if (unflushedProgress > 0) {
      _actions.default.activityTick({
        id,
        increment: unflushedProgress
      });

      unflushedProgress = 0;
    }

    lastUpdateTime = t;
  };

  return {
    start() {
      _actions.default.startActivity({
        id,
        text,
        type: _constants.ActivityTypes.Progress,
        current: start,
        total
      });
    },

    setStatus(statusText) {
      _actions.default.setActivityStatusText({
        id,
        statusText
      });
    },

    tick(increment = 1) {
      unflushedProgress += increment; // Have to manually track this :/

      updateProgress();
    },

    panicOnBuild(errorMeta, error) {
      span.finish();

      _actions.default.setActivityErrored({
        id
      });

      return reporter.panicOnBuild(errorMeta, error);
    },

    panic(errorMeta, error) {
      span.finish();

      _actions.default.endActivity({
        id,
        status: _constants.ActivityStatuses.Failed
      });

      return reporter.panic(errorMeta, error);
    },

    done() {
      updateProgress(true);
      span.finish();

      _actions.default.endActivity({
        id,
        status: _constants.ActivityStatuses.Success
      });
    },

    set total(value) {
      unflushedTotal = value;
      updateProgress();
    },

    span
  };
};

exports.createProgressReporter = createProgressReporter;
