// gatsby-cli/lib/reporter/reporter-timer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.createTimerReporter = void 0;

var _actions = _interopRequireDefault(require("./redux/actions"));

var _constants = require("./constants");

/*
 * This module is used when calling reporter.
 * these logs
 */
const createTimerReporter = ({
  text,
  id,
  span,
  reporter
}) => {
  return {
    start() {
      _actions.default.startActivity({
        id,
        text,
        type: _constants.ActivityTypes.Spinner
      });
    },

    setStatus(statusText) {
      _actions.default.setActivityStatusText({
        id,
        statusText
      });
    },

    panicOnBuild(errorMeta, error) {
      span.finish();

      _actions.default.setActivityErrored({
        id
      });

      return reporter.panicOnBuild(errorMeta, error);
    },

    panic(errorMeta, error) {
      span.finish();

      _actions.default.endActivity({
        id,
        status: _constants.ActivityStatuses.Failed
      });

      return reporter.panic(errorMeta, error);
    },

    end() {
      span.finish();

      _actions.default.endActivity({
        id,
        status: _constants.ActivityStatuses.Success
      });
    },

    span
  };
};

exports.createTimerReporter = createTimerReporter;
